Instructions for cross-compiling ffi packages

Hi,
These instructions worked nicely, thanks - I'm impressed how things just work!
However, it doesn't cover using packages with foreign dependencies; in this case I was trying to use one of the sqlite3 crates. I'd be interested in some hints for making those work if you have any! 

Thanks again for the very useful guide!

@Ogeon I'm having problems cross-compiling my Rust project now, also because of openssl. 

I'm targeting a Pi running Raspbian, and I'm using a local 32 bit VM with Debian Jessie for cross-compiling. All the steps in your guide up until executing cross32 were successful. Here are my Cargo.toml dependencies:

```
websocket = "0.12.2"
tiny_http = "0.4.0"
protobuf = "1.0.2"
time = "0.1.32"
```

First I got some missing header files:

```
/home/vagrant/.cargo/registry/src/github.com-121aea75f9ef2ce2/openssl-sys-0.6.5/src/openssl_shim.c:1:26: fatal error: openssl/hmac.h: No such file or directory
```

I got around that by doing `export OPENSSL_INCLUDE_DIR=/usr/include` - and I probably did a `apt-get install libssl-dev` first.

Next up I got 

```
/usr/include/openssl/hmac.h:61:33: fatal error: openssl/opensslconf.h: No such file or directory
```

I got around that with this one:

```
ln -s /usr/include/i386-linux-gnu/openssl/opensslconf.h /usr/include/openssl
```

But now I immediately get a whole bunch of compilation problems when running `cross32 build`, for example:

<pre>
/usr/include/stdlib.h:951:12: error: storage class specified for parameter ‘getloadavg’
 extern int getloadavg (double __loadavg[], int __nelem)
            ^
/usr/include/stdlib.h:952:6: warning: ‘__nothrow__’ attribute ignored [-Wattributes]
      __THROW __nonnull ((1));
      ^
/usr/include/stdlib.h:955:31: fatal error: bits/stdlib-float.h: No such file or directory
 #include <bits/stdlib-float.h>
                               ^
compilation terminated.
thread '<main>' panicked at 'explicit panic', /home/vagrant/.cargo/registry/src/github.com-121aea75f9ef2ce2/gcc-0.3.13/src/lib.rs:510
</pre>

 If I understand correctly, the thing to do is:
- install libssl-dev on my Pi
- Find what files were installed on the Pi using something like `dpkg --listfiles libssl-dev`
- Copy those files to the custom sysroot on my cross-compiler VM, i.e. under `~/pi-tools/arm-bcm2708/arm-bcm2708hardfp-linux-gnueabi/arm-bcm2708hardfp-linux-gnueabi/sysroot`
- unset the `OPENSSL_INCLUDE_DIR`, or at least point it inside the above sysroot

Am I on the right track here?

--- a/README.md
+++ b/README.md
@@ -34,7 +34,8 @@ git clone https://github.com/raspberrypi/tools.git ~/pi-tools
 
 The next step is to compile the Rust compiler and standard libraries. The
 standard libraries are particularly important, since they have to be compiled
-for the ARM platform to make them usable in your program.
+for the ARM platform to make them usable in your program. See Appendix A
+for information on how to add more libraries.
 
 start by cloning the Rust repository and `cd` into it:
 
@@ -124,7 +125,8 @@ built for ARM. There is currently no good way to tell Cargo to tell `rustc` to
 tell `gcc` to tell `ld` to look somewhere else, so we have to do it for them.
 
 The Raspberry Pi toolchain contains a directory with various system
-directories filled with common libraries. This is where we want `ld` to look
+directories filled with common libraries (if you want to add more libraries,
+see Appendix A). This is where we want `ld` to look
 for things, so we are going to use a simple script to tell `gcc` to tell it
 where this directory can be found. Create a file in the binary directory you
 added to your `$PATH` before
@@ -194,3 +196,46 @@ That's it! You should now be ready to cross compile your Raspberry Pi
 projects.
 
 Have fun!
+
+
+# Appendix A: Extending the toolset to support more system dependencies
+
+Let's say your project uses some crate that depends on having openssl
+installed on the system. In this case you have to install the package
+manually into the ARM toolset.
+
+Get these packages either from the raspberry, or download them online.
+
+We'll assume that you're running Raspbian (i.e. deb files), but it should be
+straight forward to adapt the steps to your dist/packages.
+
+If you do `apt-cache show libssl1.0.0` on the raspberry, you'll see this in the
+output:
+
+    Filename:    pool/main/o/openssl/libssl1.0.0_1.0.1e-2+rvt+deb7u17_armhf.deb
+
+You should be able to find a match for that under ftp.debian.org/debian/pool, so
+the resulting URL in this case is
+
+    http://ftp.debian.org/debian/pool/main/o/openssl/libssl-dev_1.0.1e-2+deb7u17_armhf.deb
+
+If it's not there, see if it is still on the raspberry under
+`/var/cache/apt/archive`.
+
+If you still can't find it, try searching for the filename online.
+
+When you have the dependencies downloaded, extract them into the ARM toolchain:
+
+    # cd into the sysroot of the arm toolchain
+    cd ~/pi-tools/arm-bcm2708/arm-bcm2708hardfp-linux-gnueabi/arm-bcm2708hardfp-linux-gnueabi/sysroot/
+
+    # move the deb files here and use `ar` to extract the contents:
+    ar p libssl1.0.0_1.0.1e-2+rvt+deb7u17_armhf.deb data.tar.gz | tar zx
+
+    # Repeat for any other dependencies you may have..
+    ar p libssl-dev_1.0.1e-2+rvt+deb7u17_armhf.deb data.tar.gz | tar zx
+    ar p zlib1g_1.2.7.dfsg-13_armhf.deb data.tar.gz | tar zx
+    ar p zlib1g-dev_1.2.7.dfsg-13_armhf.deb data.tar.gz | tar zx
+
+Now you're ready to build the project.
+

